### PR TITLE
add: mutating webhook

### PR DESCRIPTION
**Pull request by VanillaSpoon**
# Issue link
https://issues.redhat.com/browse/RHOAIENG-1991

# What changes have been made
Add Mutating Webhook to add OAuth SideCar to RayCluster

# Verification steps
Setup environment.
Build and Deploy the codeflare-operator

Ensure all configurations have been applied

From the sdk, create a raycluster
- Ensure the raycluster is deployed correctly
- Inspect the raycluster yaml to enusre the oauth configuraitons have been applied
- Inspect the raycluster head node and ensure the oauth is running alongside it.

## Checks
- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

File: config/default/webhookcainjection_patch.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    app.kubernetes.io/name: mutatingwebhookconfiguration
+    app.kubernetes.io/instance: mutating-webhook-configuration
+    app.kubernetes.io/component: webhook
+    app.kubernetes.io/created-by: codeflare-operator
+    app.kubernetes.io/part-of: codeflare-operator
+    app.kubernetes.io/managed-by: kustomize
+  name: mutating-webhook-configuration
+  annotations:
+     service.beta.openshift.io/inject-cabundle: "true"
+     service.beta.openshift.io/serving-cert-secret-name: codeflare-operator-raycluster-webhook-cert
```
Comment (astefanutti) [inline config/default/webhookcainjection_patch.yaml:15]: Is it really needed?

File: config/certmanager/kustomizeconfig.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# This configuration is for teaching kustomize how to update name ref and var substitution
```
Comment (astefanutti) [inline config/certmanager/kustomizeconfig.yaml:1]: Can that file be removed? I don't think that's needed.

File: config/certmanager/kustomization.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+resources:
```
Comment (astefanutti) [inline config/certmanager/kustomization.yaml:1]: The `certmanager` directory can be removed cert-manager is not used.

File: PROJECT [hunk 4]
```diff
@@ -15,5 +15,9 @@ resources:
   domain: codeflare.dev
```
Comment (astefanutti) [inline PROJECT:15]: Probably the domain needs to be fixed.

File: config/default/webhookcainjection_patch.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# This patch add annotation to admission webhook config
+apiVersion: admissionregistration.k8s.io/v1
+kind: MutatingWebhookConfiguration
+metadata:
+  labels:
+    app.kubernetes.io/name: mutatingwebhookconfiguration
```
Comment (astefanutti) [inline config/default/webhookcainjection_patch.yaml:6]: Those labels can be removed as there are added using common labels.

File: config/webhook/service.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
```
Comment (astefanutti) [inline config/webhook/service.yaml:4]: Those labels can be removed.

File: pkg/controllers/raycluster_webhook.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var rayclusterlog = logf.Log.WithName("raycluster-resource")
+
+func (r *RayClusterDefaulter) SetupWebhookWithManager(mgr ctrl.Manager) error {
```
Comment (astefanutti) [inline pkg/controllers/raycluster_webhook.go:32]: The convention is to have this method a function that instantiate the defaulter.

Comment (ChristianZaccaria) [reply pkg/controllers/raycluster_webhook.go:32]: I am creating and adding a validating webhook, this is what I have working so far, might be what we're looking for. Correct me if I'm wrong:
https://gist.github.com/ChristianZaccaria/58eaace628ac6a07bf058fac67ab2222 

TLDR:
```
func (w *RayClusterWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
	return ctrl.NewWebhookManagedBy(mgr).
		For(&rayv1.RayCluster{}).
		WithDefaulter(&RayClusterDefaulter{}).
		WithValidator(&RayClusterValidator{}).
		Complete()
}
```

File: config/default/webhookcainjection_patch.yaml (hunk 1 above)
Comment (astefanutti) [reply config/default/webhookcainjection_patch.yaml:15]: Yes, I meant to remove only `service.beta.openshift.io/serving-cert-secret-name`.

File: config/default/webhookcainjection_patch.yaml (hunk 5 above)
Comment (VanillaSpoon) [reply config/default/webhookcainjection_patch.yaml:6]: These were generated, so I was unsure about removing them

File: pkg/controllers/raycluster_webhook.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+				rayclusterlog.Info("OAuth sidecar already exists, no patch needed")
+				alreadyExists = true
+				break // exits the for loop
+			}
+		}
+
+		if !alreadyExists {
+			rayclusterlog.Info("Adding OAuth sidecar container")
+			// definition of the new container
+			newOAuthSidecar := corev1.Container{
```
Comment (sutaakar) [inline pkg/controllers/raycluster_webhook.go:74]: Would it have sense to limit Resources for oauth container?

File: config/webhook/service.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: webhook-service
+  namespace: openshift-operators
```
Comment (sutaakar) [inline config/webhook/service.yaml:5]: Shouldn't this be `      namespace: system`?

File: pkg/controllers/raycluster_webhook.go (hunk 8 above)
Comment (VanillaSpoon) [reply pkg/controllers/raycluster_webhook.go:74]: Thanks @sutaakar 
I feel this is a good idea, I'd be unsure of the limits to set, so I'll pull in @astefanutti  :)
